##  Tasca M3 T02

### Exercici 1
***

Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [1]:
import pandas as pd

data = pd.read_csv("DelayedFlights.csv")

df = pd.DataFrame(data)

# Observació de les dades que conté el dataset

print (df.head(10))

   Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   
5           6  2008      1           3          4   1937.0        1830   
6          10  2008      1           3          4    706.0         700   
7          11  2008      1           3          4   1644.0        1510   
8          15  2008      1           3          4   1029.0        1020   
9          16  2008      1           3          4   1452.0        1425   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...

In [2]:
# Observació de quin tipus de dades hi ha al dataset

print (df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

És un dataset extens (1936758 files × 30 columnes) que conté informació sobre els vols durant el 2008. Les variables del dataset són les següents:

- **Year**: L'any del vol.

- **Month**: El mes del vol.

- **DayofMonth**: El dia del mes del vol.

- **DayOfWeek**: El dia de la setmana del vol.

- **DepTime**: L'hora de sortida real del vol.

- **CRSDepTime**: L'hora de sortida programada del vol.

- **ArrTime**: L'hora d'arribada real del vol.

- **CRSArrTime**: L'hora d'arribada programada del vol.

- **UniqueCarrier**: El codi d'identificació de l'aerolínia.

- **FlightNum**: El número de vol.

- **TailNum**: El número de cua de l'aeronau.

- **ActualElapsedTime**: Durada real del vol (minuts).

- **CRSElapsedTime**: Durada del vol programat (minuts).

- **AirTime**: El temps de vol en l'aire (minuts).

- **ArrDelay**: El retard en l'arribada (minuts).

- **DepDelay**: El retard en la sortida (minuts).

- **Origin**: L'aeroport d'origen del vol.

- **Dest**: L'aeroport de destinació del vol.

- **Distance**: La distància del vol (milles).

- **TaxiIn**: El temps de rodatge en l'arribada (minuts).

- **TaxiOut**: El temps de rodatge en la sortida (minuts).

- **Cancelled**: Vols cancel·lats (Si(1) o No(0)).

- **CancellationCode**: El codi de cancel·lació del vol.

- **Diverted**: Vols desviats a un altre aeroport (Si(1) o No(0)).

- **CarrierDelay**: El temps de retard degut a l'aerolínia (minuts).

- **WeatherDelay**: El temps de retard degut al les condicions climàtiques (minuts).

- **NASDelay**: El temps de retard degut al Sistema Nacional d'Aviació (minuts).

- **SecurityDelay**: El temps de retard degut a problemes de seguretat (minuts).

- **LateAircraftDelay**: El temps de retard degut a problemes amb l'avió (minuts).


De moment totes les columnes poden aportar informació rellevant per fer un anàlisi d'aquest dataset. Més endavant, segons les necessitats, s'escollirà quines són les més adecuades. Seguidament, mirem si hi ha valors nuls per poder tratar les dades i decidir quines són les importants i/o les que necessitem.

In [3]:
# Observar si hi ha valors nuls al dataset

print (df.isnull().sum())


Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64


Les variables 'ArrTime', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay', 'TaxiIn' i 'TaxiOut' tenen valors nuls, però proporcionalment al total de registres (casi 2M), no sembla que influeixin. Per tant, de moment no les eliminarem, ni modificarem. En canvi, les variables 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay' i 'LateAircraftDelay' si que tenen un nombre considerable de nuls (un terç del total de registres). Procedim a analitzar aquest fet:

In [4]:
# Mirar si la falta de registres coincideix amb els vols cancelats, desviats i els que no es consideren delays (< 15 min)
# (si no són delays és lògic que no tinguin registres en les variables sobre les causes d'aquests)


df_arr = df[df['ArrDelay'] < 15].shape[0]

df_can = df['Cancelled'].value_counts()

df_des = df['Diverted'].value_counts()

suma = df_arr + df_can[1] + df_des[1]

print ('Vols sense Delay de més de 15 min:', df_arr)
print ('Vols cancelats:', df_can[1])
print ('Vols desviats:', df_des[1])
print ('Suma total:', suma)

Vols sense Delay de més de 15 min: 680883
Vols cancelats: 633
Vols desviats: 7754
Suma total: 689270


Coincideixen els números (la suma total i els valors nuls), per tant, procedim a canviar els valors nuls per un 0.

In [2]:
# Canvi dels valors nuls a 0

df_2 = ['CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']

df[df_2] = df[['CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].fillna(0)

print (df.isnull().sum())

Unnamed: 0              0
Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
DepTime                 0
CRSDepTime              0
ArrTime              7110
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum                 5
ActualElapsedTime    8387
CRSElapsedTime        198
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
TaxiIn               7110
TaxiOut               455
Cancelled               0
CancellationCode        0
Diverted                0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
dtype: int64


In [3]:
# Eliminació dels valors nuls (com farem una selecció aleatòria de 200.000 registres, 
# podem eliminar-los sense influir en l'anàlisi del dataset)

df = df.dropna()

print (df.isnull().sum())

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64


Ara ja tenim el dataset que necessitem. Seguidament, farem la reducció a 200.000 registres aleatoris.

In [4]:
# Reducció de la dimensió del dataset de manera aleatòria en 200.000 registres

df = df.sample(n=200000)

print (df)

         Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1614628     5879594  2008     10          14          2   1324.0        1310   
1178375     3936138  2008      7          31          4   1445.0        1403   
637785      2036082  2008      4          14          1   2218.0        2120   
264982       841780  2008      2           9          6   1824.0        1315   
107833       344972  2008      1          17          4   1957.0        1835   
...             ...   ...    ...         ...        ...      ...         ...   
845972      2869791  2008      5          12          1   1525.0        1320   
1409033     4785774  2008      8          11          1   1451.0         935   
774291      2545562  2008      5          12          1   1055.0        1045   
775878      2552894  2008      5          25          7   1125.0        1108   
815080      2727410  2008      5           7          3   1623.0        1615   

         ArrTime  CRSArrTime UniqueCarr

Visulaitzant les dades, podem determinar que les següents columnes no fa falta seleccionar-les:

- Year: tots són del mateix any.
- FlightNum i TailNum: no ens demanen informació tan específica (encara que en el següent exercici es pot utilitzar el FlightNum en comptes de l'origen i destí).
- CarrierDelay, WeatherDelay, NASDelay, SecurityDelay i LateAircraftDelay: no necessitem saber les causes dels delays.
- Cancelled i Diverted: en el següent exercici veiem que no hi ha vols cancel·lats o desviats.
- CRSDepTime, CRSArrTime i CRSElapsedTime: és informació redundant, ja que amb el DepTime i DepDelay podem extreure el CRSDepTime (passa el mateix amb les altres variables).



In [5]:
df = df.drop(['Year', 'FlightNum', 'TailNum', 
              'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay',
              'Cancelled', 'Diverted', 'CancellationCode',
              'CRSDepTime', 'CRSArrTime', 'CRSElapsedTime'], axis=1)

print (df)

         Unnamed: 0  Month  DayofMonth  DayOfWeek  DepTime  ArrTime  \
1614628     5879594     10          14          2   1324.0   1418.0   
1178375     3936138      7          31          4   1445.0   1620.0   
637785      2036082      4          14          1   2218.0   2332.0   
264982       841780      2           9          6   1824.0     26.0   
107833       344972      1          17          4   1957.0   2203.0   
...             ...    ...         ...        ...      ...      ...   
845972      2869791      5          12          1   1525.0   2013.0   
1409033     4785774      8          11          1   1451.0   1831.0   
774291      2545562      5          12          1   1055.0   1210.0   
775878      2552894      5          25          7   1125.0   1619.0   
815080      2727410      5           7          3   1623.0   1942.0   

        UniqueCarrier  ActualElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
1614628            AS               54.0     34.0      20.0      14

### Exercici 2
***

Fes un informe complet del dataset:

- Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
- Troba quantes dades faltants hi ha per columna.
- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
- Fes una taula de les aerolínies amb més endarreriments acumulats.
- Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards. Aporta allò que consideris rellevant.


In [6]:
# Resum estadístic del dataset

print (df.describe().round(2))

       Unnamed: 0      Month  DayofMonth  DayOfWeek    DepTime    ArrTime  \
count   200000.00  200000.00   200000.00  200000.00  200000.00  200000.00   
mean   3337740.78       6.10       15.76       3.98    1518.94    1610.05   
std    2066138.26       3.48        8.76       2.00     450.27     548.49   
min          1.00       1.00        1.00       1.00       1.00       1.00   
25%    1514285.00       3.00        8.00       2.00    1203.00    1315.00   
50%    3236757.50       6.00       16.00       4.00    1545.00    1716.00   
75%    4972739.00       9.00       23.00       6.00    1900.00    2030.00   
max    7009664.00      12.00       31.00       7.00    2400.00    2400.00   

       ActualElapsedTime    AirTime   ArrDelay   DepDelay   Distance  \
count          200000.00  200000.00  200000.00  200000.00  200000.00   
mean              133.37     108.33      42.26      43.12     764.84   
std                72.15      68.73      56.52      52.96     574.05   
min               

La mitjana del mes és aproximadament 6.10, la qual cosa suggereix que els vols en aquest conjunt de dades estan distribuïts al llarg de l'any. Tanmateix, la mitjana del dia del mes (mean = 15.76) i del dia de la setmana (mean = 3.98) indica que els vols en el conjunt de dades es distribueixen al llarg del mes i de la seman per igual.

El temps dels vols té una mitjana d'aproximadament 108.33 minuts i la La distancia mitjana dels vols és d'aproximadament 764.84 milles. Les mitjanes dels retards en l'arribada i la sortida són d'al voltant de 42.26 i 43.12 minuts, respectivament. Això suggereix que, en mitjana, els vols tendeixen a arribar una mica abans de l'hora programada d'arribada, però també a enlairar amb retard.

En resum, aquesta anàlisi estadística proporciona una visió molt general de les característiques del conjunt de dades de vols, incloent-hi detalls sobre els temps programats i reals, les dates i els retards.

In [7]:
# Dades faltants per columna

print (df.isnull().sum())

Unnamed: 0           0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
ArrTime              0
UniqueCarrier        0
ActualElapsedTime    0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
dtype: int64


Desprès de la neteja que hem fet al exercici anterior, no queda cap valor nul en el dataset.

In [8]:
# Creació de columnes noves

# Velocitat mitjana de vol en km/h

df['vel_mitj_vol'] = ((df['Distance'] / 1.609) / (df['AirTime'] / 60)).round(2)

print (df.head(10))


         Unnamed: 0  Month  DayofMonth  DayOfWeek  DepTime  ArrTime  \
1614628     5879594     10          14          2   1324.0   1418.0   
1178375     3936138      7          31          4   1445.0   1620.0   
637785      2036082      4          14          1   2218.0   2332.0   
264982       841780      2           9          6   1824.0     26.0   
107833       344972      1          17          4   1957.0   2203.0   
954903      3208857      6          27          5    914.0   1246.0   
609011      1917122      4          24          4   1316.0   1442.0   
1113062     3706948      7          16          3   1138.0   1433.0   
1557569     5547093     10          27          1   1659.0   2052.0   
1740025     6494728     12          12          5   1032.0   1211.0   

        UniqueCarrier  ActualElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
1614628            AS               54.0     34.0      20.0      14.0    CDV   
1178375            EV               95.0     64.0      50.

In [11]:
# Si el vol aha arribat tard o no (15 min > ArrDelay)

df['Arriba_tard'] = (df['ArrDelay'] > 15).astype(int)

print (df.head(10))
print ('')
print ('Arriba_tard:')
print ('1 si ha arribat tard')
print ('0 si no ha arribat tard')

         Unnamed: 0  Month  DayofMonth  DayOfWeek  DepTime  ArrTime  \
1614628     5879594     10          14          2   1324.0   1418.0   
1178375     3936138      7          31          4   1445.0   1620.0   
637785      2036082      4          14          1   2218.0   2332.0   
264982       841780      2           9          6   1824.0     26.0   
107833       344972      1          17          4   1957.0   2203.0   
954903      3208857      6          27          5    914.0   1246.0   
609011      1917122      4          24          4   1316.0   1442.0   
1113062     3706948      7          16          3   1138.0   1433.0   
1557569     5547093     10          27          1   1659.0   2052.0   
1740025     6494728     12          12          5   1032.0   1211.0   

        UniqueCarrier  ActualElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
1614628            AS               54.0     34.0      20.0      14.0    CDV   
1178375            EV               95.0     64.0      50.

In [12]:
# Taula de les aerolínies amb més endarreriments acumulats

df_aero = df

delays = df_aero.groupby('UniqueCarrier')['ArrDelay'].sum().reset_index()

delays.rename(columns={'ArrDelay': 'TotalDelays'}, inplace=True)

aero_delays = delays.sort_values(by='TotalDelays', ascending=False)


print(aero_delays)


   UniqueCarrier  TotalDelays
17            WN    1178095.0
1             AA     923759.0
15            UA     686149.0
11            MQ     669476.0
14            OO     621967.0
18            XE     535948.0
6             DL     473601.0
5             CO     425203.0
7             EV     409466.0
19            YV     379586.0
16            US     365688.0
12            NW     359675.0
4             B6     315717.0
9             FL     312981.0
13            OH     288629.0
0             9E     250594.0
3             AS     146593.0
8             F9      80433.0
10            HA      27246.0
2             AQ       1484.0


La companyia amb més minuts de delays acumuats es la WN.

In [16]:
# Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards


# Vols més llargs (rutes més llargues)

df_vols = df

vols_llargs = df_vols[['Origin', 'Dest', 'Distance']].sort_values(by='Distance', ascending=False)

print(vols_llargs)

print ('Els vols més llargs són els de la ruta EWR-HNL / HNL-EWR')
print (' ')

# Vols més endarrerits

df_vols2 = df

vols_enda = df_vols2.groupby(['Origin', 'Dest'])['ArrDelay'].sum().reset_index()
                             
vols_enda.rename(columns={'ArrDelay': 'Vols_mes_dealy'}, inplace=True)

vols_mes = vols_enda.sort_values(by='Vols_mes_dealy', ascending=False)

print (vols_mes)
print ('Els vols amb més delay són els que es mostren en pantalla')

        Origin Dest  Distance
875809     HNL  EWR      4962
871328     EWR  HNL      4962
365506     EWR  HNL      4962
721700     EWR  HNL      4962
870877     EWR  HNL      4962
...        ...  ...       ...
865204     PSG  WRG        31
864305     PSG  WRG        31
1240809    WRG  PSG        31
47146      IAD  DCA        24
1310766    JFK  LGA        11

[200000 rows x 3 columns]
Els vols més llargs són els de la ruta EWR-HNL / HNL-EWR
 
     Origin Dest  Vols_mes_dealy
3470    ORD  LGA         27752.0
3440    ORD  EWR         21954.0
2507    LAX  SFO         21916.0
1641    EWR  ORD         21434.0
2580    LGA  ORD         20526.0
...     ...  ...             ...
2482    LAX  MEM           -21.0
2357    LAS  BMI           -31.0
4500    SMF  OGG           -33.0
3300    OAK  TUS           -51.0
1213    DEN  LIH           -67.0

[4826 rows x 3 columns]
Els vols amb més delay són els que es mostren en pantalla


### Exercici 3
***

Exporta el dataset net i amb les noves columnes a Excel.

In [17]:
df.to_excel('Vols_Delays_2008.xlsx')